src/linux/ebpf.cpp

```diff
@@ -14,4 +14,99 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "linux/ebpf.hpp"
\ No newline at end of file
+#include "linux/ebpf.hpp"
+
+#include <errno.h>
+#include <linux/bpf.h>
+#include <sys/syscall.h>
+
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "stout/foreach.hpp"
+#include "stout/os.hpp"
+
+using std::ostream;
+using std::string;
+using std::vector;
+
+namespace ebpf {
+
+namespace internal {
+
+// Wrapper around the `bpf` syscall.
+Try<int> syscall_bpf(int cmd, union bpf_attr* attr, size_t size)
+{
+  int result = syscall(__NR_bpf, cmd, attr, size);
+  if (result == -1) {
+    return Error(
+      "BFP syscall failed with error '" + string(std::strerror(errno)) + "'");
+  }
+  return result;
+}
+
+
+uint64_t pointer_to_int(const void* pointer)
+{
+  return (uint64_t)(unsigned long)pointer;
+}
+
+} // namespace internal {
+
+ProgramBuilder::ProgramBuilder(enum bpf_prog_type _type) : type(_type) {}
+
+
+void ProgramBuilder::append(std::vector<struct bpf_insn> instructions)
+{
+  program.insert(program.end(), instructions.begin(), instructions.end());
+}
+
+
+size_t ProgramBuilder::instructions() const { return program.size(); }
+
+
+Program ProgramBuilder::build()
+{
+  union bpf_attr attribute;
+  std::memset(&attribute, 0, sizeof(attribute));
+
+  attribute.prog_type = type;
+  attribute.insn_cnt = instructions();
+  attribute.insns = internal::pointer_to_int(program.data());
+  attribute.license = internal::pointer_to_int("Apache 2.0");
+
+  return Program{.attribute = attribute};
+}
+
+
+Try<int> load(const Program& program, bool logErrors)
+{
+  union bpf_attr attribute = program.attribute;
+  Try<int> fd = 0; // placeholder because Try<int>() DNE
+  string logs = "";
+
+  if (logErrors) {
+    const int LOG_SIZE = 8196;
+    const int LOG_LEVEL = 1;
+
+    char* log_buffer = (char*)malloc(LOG_SIZE);
```

Comment:
this is a memory leak! we could stack (or unique_ptr + new) allocate it inside the if EACCESS if condition:

```cpp
char verifier_log_buffer[VERIFIER_LOG_SIZE];

// Ensure the log is null terminated if the verifier doesn't write anything.
verifier_log_buffer[0] = '\0';

attribute.log_buf = static_cast<uint64_t>(verifier_log_buffer);

// We don't bother looking for ENOSPC because we use a very large buffer size.
XXX

// Maybe check that it's EACCESS again..?
ErrnoError error;
CHECK_EQ(EACCESS, error.errno) << "Expected verifier error again, got: " << error;

return Error("bpf verifier rejected the program: " + string(verifier_log_buffer))
```